### PR TITLE
C10 library for shared functionality between PyTorch and Caffe2

### PR DESCRIPTION
**Pull request by ezyang**
Stacked on https://github.com/pytorch/pytorch/pull/7275

Requested by @orionr 

```
    Create c10 library for shared functionality between PyTorch and Caffe2.
        
    The c10 library is used by both PyTorch and Caffe2.  We start it off
    small and minimal, and we will suck in functionality from PyTorch/Caffe2
    that we want to use from the other side.
        
    To start, c10 contains:
        
    c10/Optional.h    An implementation of std::optional
    c10/Error.h       Error reporting functionality
    c10/Registry.h    The Caffe2 registry
 ```

### CODE REVIEW

**Comment by ezyang** [comment]
CC @smessmer @dzhulgakov

File: aten/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ else()
   cmake_dependent_option(
       NO_MKLDNN "Legacy no MKLDNN" OFF
       "USE_MKLDNN" ON)
+  add_subdirectory(../c10 c10)
```
Comment (smessmer) [inline aten/CMakeLists.txt]: Seems weird to me to include stuff from parent directory. Shouldn't the parent directory do `add_subdirectory(c10)` instead?

File: aten/src/ATen/CMakeLists.txt [hunk 2]
```diff
@@ -242,7 +242,9 @@ ENDIF()
 # of NO_CUDA (it always builds the same way); libATen_cuda.so is only
 # built when NO_CUDA=0 and CUDA is available.
 
-add_library(ATen_cpu ${AT_LINK_STYLE} ${all_cpu_cpp})
+add_library(ATen_cpu ${AT_LINK_STYLE} ${all_cpu_cpp} ${C10_SRCS})
```
Comment (smessmer) [inline aten/src/ATen/CMakeLists.txt]: Why not `target_link_libraries` statically to c10 instead?

File: c10/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+project(c10 CXX)
+
+include(CheckCXXSourceCompiles)
+
+check_cxx_source_compiles("#include <execinfo.h>
```
Comment (smessmer) [inline c10/CMakeLists.txt]: This might be better with `FindBacktrace.cmake`

File: caffe2/CMakeLists.txt [hunk 4]
```diff
@@ -131,7 +131,8 @@ target_include_directories(caffe2_protos INTERFACE $<INSTALL_INTERFACE:include>)
 target_link_libraries(caffe2_protos PUBLIC protobuf::libprotobuf)
 
 # Compile exposed libraries.
-add_library(caffe2 ${Caffe2_CPU_SRCS})
+add_library(caffe2 ${Caffe2_CPU_SRCS} ${C10_SRCS})
```
Comment (smessmer) [inline caffe2/CMakeLists.txt]: same here. Why not link?

File: caffe2/core/registry.h [hunk 5]
```diff
@@ -1,213 +1,13 @@
...
 
-#include <algorithm>
-#include <cstdio>
-#include <cstdlib>
-#include <functional>
-#include <memory>
-#include <mutex>
+#include <c10/Registry.h>
 
+// Kept for BC reasons
```
Comment (smessmer) [inline caffe2/core/registry.h]: Which BC would break if we remove them? We should avoid dependency cycles.

**Comment by smessmer** [comment]
Also, can you add me as a CODEOWNER ?

File: caffe2/core/registry.h (hunk 5 above)
Comment (smessmer) [reply caffe2/core/registry.h]: nvm, I thought this is c10/registry.h

**Comment by ezyang** [comment]
We're not going to take this PR as it is, but I'm keeping it open because it contains some important fixes which I need to backport.

**Comment by ezyang** [comment]
Obsoleted by @Yangqing's newer PRs.

**Pull request closed without merging**